server/pulp/server/async/app.py

```diff
@@ -49,8 +75,72 @@ def initialize_worker(sender, instance, **kwargs):
     """
     initialization.initialize()
 
+    # Delete any potential old state
     tasks._delete_worker(sender, normal_shutdown=True)
 
     # Create a new working directory for worker that is starting now
     common_utils.delete_worker_working_directory(sender)
     common_utils.create_worker_working_directory(sender)
+
+    # If the worker is a resource manager, try to acquire the lock, or wait until it
+    # can be acquired
+    if sender.startswith(constants.RESOURCE_MANAGER_WORKER_NAME):
+        get_resource_manager_lock(sender)
+
+
+def get_resource_manager_lock(sender):
+    """
+    Tries to acquire the resource manager lock. If the lock cannot be acquired immediately, it
+    will wait until the currently active instance becomes unavailable, at which point the worker
+    cleanup routine will clear the lock for us to acquire. A worker record will be created so that
+    the waiting resource manager will appear in the Status API. We override the SIGTERM signal
+    handler so that that the worker record will be immediately cleaned up if the process is killed
+    while in this states.
+    """
+    name = constants.RESOURCE_MANAGER_WORKER_NAME + "@" + platform.node()
+    lock = ResourceManagerLock(name=name)
+
+    assert name.startswith(constants.RESOURCE_MANAGER_WORKER_NAME)
+
+    with custom_sigterm_handler(name):
+        # Whether this is the first lock availability check for this instance
+        _first_check = True
+
+        while True:
+            # Create / update the worker record so that Pulp knows we exist
+            Worker.objects(name=sender).update_one(set__last_heartbeat=datetime.utcnow(),
+                                                   upsert=True)
+            try:
+                lock.save()
+
+                msg = _("Resource manager '%s' has acquired the resource manager lock" % name)
+                _logger.info(msg)
+                break
+            except mongoengine.NotUniqueError:
+                # Only log the message the first time
+                if _first_check:
+                    msg = _("Resource manager '%s' attempted to acquire the the resource manager "
+                            "lock but was unable to do so. It will retry every %d seconds until "
+                            "the lock can be acquired." % (name, name,
+                                                           constants.CELERY_CHECK_INTERVAL))
+                    _logger.info(msg)
+                    _first_check = False
+
+                time.sleep(constants.CELERY_CHECK_INTERVAL)
+
```

Comment:
@mhrivnak We already have a set of behaviors which are implicit everywhere in Pulp for just such a situation. The user can choose to have our DB auto retry by enabling an option, or if they don't an exception will be raised and all processes will exit (which is expected). Why would we need more than what we already have?